lib/inMemFileDB.js

```diff
@@ -0,0 +1,346 @@
+        }
+        client._subscribe = client._subscribe || {};
+        if (!client._subscribe[type]) client._subscribe[type] = [];
+        const s = client._subscribe[type];
+
+        if (pattern instanceof Array) {
+            pattern.forEach(pattern => {
+                for (let i = 0; i < s.length; i++) {
+                    if (s[i].pattern === pattern) return;
+                }
+
+                s.push({pattern: pattern, regex: new RegExp(tools.pattern2RegEx(pattern)), options: options});
+            });
+        } else {
+            for (let i = 0; i < s.length; i++) {
+                if (s[i].pattern === pattern) {
+                    typeof cb === 'function' && cb();
+                    return;
+                }
+            }
+
+            s.push({pattern: pattern, regex: new RegExp(tools.pattern2RegEx(pattern)), options: options});
+        }
+        typeof cb === 'function' && cb();
+    }
+
+    handleUnsubscribe(client, type, pattern, cb) {
+        if (!client._subscribe || !client._subscribe[type]) {
+            if (typeof cb === 'function') cb();
+            return;
+        }
+        const s = client._subscribe[type];
+        if (pattern instanceof Array) {
+            pattern.forEach(pattern => {
+                for (let i = 0; i < s.length; i++) {
+                    if (s[i].pattern === pattern) {
+                        s.splice(i, 1);
+                        return;
+                    }
+                }
+            });
+        } else {
+            for (let i = 0; i < s.length; i++) {
+                if (s[i].pattern === pattern) {
+                    s.splice(i, 1);
+                    if (typeof cb === 'function') cb();
+                    return;
+                }
+            }
+        }
+        if (typeof cb === 'function') cb();
+
+    }
+
+    publishToClients(client, type, id, obj) {
+        throw new Error('no communication handling implemented');
+    }
+
+    deleteOldBackupFiles() {
+        // delete files only if settings.backupNumber is not 0
+        const files = fs.readdirSync(this.backupDir);
+        files.sort();
+        const limit = Date.now() - this.settings.backup.hours * 3600000;
+
+        for (let f = files.length - 1; f >= 0; f--) {
+            if (!files[f].endsWith(this.settings.fileDB.fileName + '.gz')) {
+                files.splice(f, 1);
+            }
+        }
+
+        while (files.length > this.settings.backup.files) {
+            const file = files.shift();
+            // extract time
+            const ms = new Date(file.substring(0, 10) + ' ' + file.substring(11, 16).replace('-', ':') + ':00').getTime();
+            if (limit > ms) {
+                try {
+                    fs.unlinkSync(this.backupDir + file);
+                } catch (e) {
+                    this.log.error(`Cannot delete file "${this.backupDir + file}: ${JSON.stringify(e)}`);
+                }
+            }
+        }
+    }
+
+    getTimeStr(date) {
```

Comment:
Isn't that defined similarly somewhere in formatDate?

Reply:
but formatDate is adapter class only ...